async-server/src/main.rs

```diff
@@ -12,22 +12,23 @@ use lines::Lines;
 use protocol::{Reply, Request};
 use romio::{TcpListener, TcpStream};
 use std::collections::HashMap;
+use std::collections::HashSet;
 use std::io;
 use std::marker::Unpin;
 use std::net::SocketAddr;
 use std::str::FromStr;
 use std::sync::Arc;
 
 /// A connection to a client, to which we can write serialized `Reply` objects.
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 struct Outbound(Arc<Mutex<Box<dyn 'static + AsyncWrite + Send + Unpin>>>);
 
-#[derive(Default)]
+#[derive(Default, Debug)]
 struct Channel {
-    subscribers: Vec<Outbound>,
+    subscribers: HashMap<SocketAddr, Outbound>,
```

Comment:
Instead of `SocketAddr`, let's use `Arc::ptr_eq` to decide whether two `Outbound` structures are the same. Then this type can remain unchanged. For a demo, I'm unconcerned with the linear scan of the vector.